File: leetcode/python/medium/Problem148.py
# Questions to ask:
# 1. What is the time complexity? O(n*log(n))
# 2. What is the space complexity? O(1)
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def merge(self, list1: ListNode | None, list2: ListNode | None):
        dummy = ListNode()
        curr = dummy

        while list1 and list2:
            if list1.val < list2.val:
                curr.next = list1
                list1 = list1.next
            else:
                curr.next = list2
                list2 = list2.next
            curr = curr.next

        if list1:
            curr.next = list1
        if list2:
            curr.next = list2

        return dummy.next

    def sortList(self, head: ListNode | None) -> ListNode | None:
        if not head or not head.next:
            return head

        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        mid = slow.next
        slow.next = None # Split it into 2 parts

        left = self.sortList(head) # first to slow
        right = self.sortList(mid) # slow.next to tail

        return self.merge(left, right)


# Problem 148
# Link: https://leetcode.com/problems/sort-list/description/
# Tips:
if __name__ == '__main__':
    s = Solution()

File: leetcode/golang/medium/Problem148.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func merge(node1 *ListNode, node2 *ListNode) *ListNode {
	dummy := &ListNode{}
	curr := dummy

	for node1 != nil && node2 != nil {
		if node1.Val < node2.Val {
			curr.Next = node1
			node1 = node1.Next
		} else {
			curr.Next = node2
			node2 = node2.Next
		}
		curr = curr.Next
	}

	if node1 != nil {
		curr.Next = node1
	}

	if node2 != nil {
		curr.Next = node2
	}

	return dummy.Next
}

func sortList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	slow := head
	fast := head.Next
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	mid := slow.Next
	slow.Next = nil

	left := sortList(head)
	right := sortList(mid)

	return merge(left, right)
}
